src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs

```diff
@@ -874,6 +875,19 @@ Task<string> IDurableOrchestrationClient.StartNewAsync<T>(string orchestratorFun
             return ((IDurableOrchestrationClient)this).StartNewAsync<T>(orchestratorFunctionName, string.Empty, input);
         }
 
+        async Task<string> IDurableOrchestrationClient.RestartAsync(string instanceId, bool restartWithNewInstanceId)
+        {
+            DurableOrchestrationStatus status = await ((IDurableOrchestrationClient)this).GetStatusAsync(instanceId, showHistory: false, showHistoryOutput: false, showInput: true);
+
+            if (status == null)
+            {
+                throw new ArgumentException($"An orchestrastion with the instanceId {instanceId} was not found.");
+            }
+
```

Comment:
We may want to do some validation of the current status. For instance, if the current status is running, then `restartWithNewInstanceId=false` will throw an exception. 

Also, what behavior do we want for instances that are still in the running state in general? Do we want to terminate the running orchestration?

Reply:
I do think it makes sense to honor the customers setting for starting orchestrators with the same instanceId.

However, whether we want to add a terminate step is a good question. I could easily add a call to TerminateAsync in this method. I can't think of a good reason to add this though, and it feels weird to force this behavior without a reason. We can choose to wait to add a potential terminate step or any other changes and see what customers think of this API and how they use it. We could release this as "in preview" until we have more feedback so if we do make changes they wont be so unexpected. What do you think?

Reply:
Eh, I think you are right that the best approach would be to not do anything automatically, and note in the C# docs that it may throw an exception if the orchestration instance is not in a state that can be restarted.

We can always harden this API later by catching any exceptions thrown by `StartNewAsync()` and give a more helpful exception message down the line.